features/storage/TESTS/kvstore/general_tests/main.cpp

```diff
@@ -0,0 +1,982 @@
+
+    kvstore->reset();
+}
+
+static void test_thread_get(const void *th_key)
+{
+    int res = kvstore->get((char *)th_key, buffer, buffer_size, &actual_size, 0);
+    TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+}
+
+//get several keys multithreaded
+static void get_several_keys_multithreaded()
+{
+    rtos::Thread kvstore_thread[num_of_threads];
+    osStatus threadStatus;
+
+    for (int i = 0; i < num_of_threads; i++) {
+        int res = kvstore->set(keys[i], data, data_size, 0);
+        TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+    }
+
+    kvstore_thread[0].start(callback(test_thread_get, "key1"));
+    kvstore_thread[1].start(callback(test_thread_get, "key2"));
+    kvstore_thread[2].start(callback(test_thread_get, "key3"));
+
+    for (int i = 0; i < num_of_threads; i++) {
+        threadStatus = kvstore_thread[i].join();
+        if (threadStatus != 0) {
+            utest_printf("\nthread %d join failed!", i + 1);
+        }
+    }
+    
+    kvstore->reset();
+}
+
+/*----------------remove()------------------*/
+
+//bad params : key is null
+static void remove_key_null()
+{
+    int res = kvstore->remove(NULL);
+    TEST_ASSERT_EQUAL(MBED_ERROR_INVALID_ARGUMENT, res);
+}
+
+//bad params : key length over key max size
+static void remove_key_length_exceeds_max()
+{
+    char key_max[KVStore::MAX_KEY_SIZE + 1] = {0};
+    memset(key_max, '*', KVStore::MAX_KEY_SIZE);
+    int res = kvstore->remove(key_max);
+    TEST_ASSERT_EQUAL(MBED_ERROR_INVALID_ARGUMENT, res);
+}
+
+//key doesn’t exist
+static void remove_non_existing_key()
+{
+    char new_key[] = "remove_key";
+    int res = kvstore->remove(new_key);
+    TEST_ASSERT_EQUAL(MBED_ERROR_ITEM_NOT_FOUND, res);
+}
+
+//key already removed
+static void remove_removed_key()
+{
+    int res = kvstore->set(key, data, data_size, 0);
+    TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+
+    res = kvstore->remove(key);
+    TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+
+    res = kvstore->remove(key);
+    TEST_ASSERT_EQUAL(MBED_ERROR_ITEM_NOT_FOUND, res);
+
+    kvstore->reset();
+}
+
+//key exist - valid flow
+static void remove_existed_key()
+{
+    int res = kvstore->set(key, data, data_size, 0);
+    TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+
+    res = kvstore->remove(key);
+    TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+
+    kvstore->reset();
+}
+
+/*----------------get_info()------------------*/
+
+KVStore::info_t info;
+
+//bad params : key is null
+static void get_info_key_null()
+{
+    int res = kvstore->get_info(NULL, &info);
+    TEST_ASSERT_EQUAL(MBED_ERROR_INVALID_ARGUMENT, res);
+}
+
+//bad params : key length over key max size
+static void get_info_key_length_exceeds_max()
+{
+    char key_max[KVStore::MAX_KEY_SIZE + 1] = {0};
+    memset(key_max, '*', KVStore::MAX_KEY_SIZE);
+    int res = kvstore->get_info(key_max, &info);
+    TEST_ASSERT_EQUAL(MBED_ERROR_INVALID_ARGUMENT, res);
+}
+
+//bad params : &info is null
+static void get_info_info_null()
+{
+    int res = kvstore->get_info(key, NULL);
```

Comment:
This should actually work, though not defined explicitly. We can use this function to check whether a key exists, therefore the info argument should defaults to NULL. Will change the base class also.

@offirko Please change this accordingly also in FileSystemStore.

Reply:
Updated my code and tests